lib/interactivity-api.php

```diff
@@ -24,9 +24,28 @@ function gutenberg_reregister_interactivity_script_modules() {
 	wp_register_script_module(
 		'@wordpress/interactivity-router',
 		gutenberg_url( '/build-module/interactivity-router/index.min.js' ),
-		array( '@wordpress/interactivity' ),
+		array(
+			array( 'id' => '@wordpress/a11y', 'import' => 'dynamic' ),
+			'@wordpress/interactivity',
+		),
 		$default_version
 	);
 }
-
 add_action( 'init', 'gutenberg_reregister_interactivity_script_modules' );
+
+function gutenberg_register_interactivity_script_module_data_hooks() {
```

Comment:
I mentioned this on https://github.com/WordPress/wordpress-develop/pull/7304 already, but:

Can't we find a better way?

This is really a step backwards from the JS I18N mechanism we implemented all those years ago 😞 

Translators loose important context because the strings are no longer co-located with the code where they're being used. This also fails as soon as soon as you are dealing with plurals.

Reply:
> I'm looking at tannin which is the lower level package that i18n relies on and it seems to be 1.49 kb gzipped.

It'a a drop-in replacement for Jed. I'm not entirely sure if you can do much about it. The same situation applies to `sprintf-js` which has similar size like `tannin`.

> This is really a step backwards from the JS I18N mechanism we implemented all those years ago 😞

By the way, it made me realize that when the user locale is English, then you don't have to pass the translation to the client. That's another aspect that is abstracted with the built-in handling for i18n and scripts.

English:

<img width="812" alt="Screenshot 2024-09-12 at 10 30 07" src="https://github.com/user-attachments/assets/b40093c8-5ea1-4a1e-908a-b1fe197aa5c6">


Polish
<img width="1045" alt="Screenshot 2024-09-12 at 10 28 23" src="https://github.com/user-attachments/assets/3f9dcb59-82e4-4110-b468-70378940dbeb">

Reply:
Tannin also includes a sprintf package: https://www.npmjs.com/package/@tannin/sprintf

> By the way, it made me realize that when the user locale is English, then you don't have to pass the translation to the client. That's another aspect that is abstracted with the built-in handling for i18n and scripts.

Yep, I've been thinking about this as a possible optimization. Some minimal i18n English-only package could be shipped like:

```js
const __ = ( text ) => text;
const _x = ( text ) => text;
const _n = ( singular, plural, number ) => number != 1 ? plural : singular;
```

Although I may be missing something there

Reply:
<img width="1260" alt="Screenshot 2024-09-12 at 12 06 12" src="https://github.com/user-attachments/assets/e5457bfb-311e-47a0-965a-343cc5fa3bb0">

It looks promising 😄

Reply:
It might even look a bit better as the drop-in replacement doesn't throw exceptions. There is only one unit test failing:

<img width="978" alt="Screenshot 2024-09-12 at 12 16 23" src="https://github.com/user-attachments/assets/6f0fb961-bdc5-4990-a979-768b6301fb37">

Which covers handling exceptions.

Reply:
We're getting a bit off topic here 😅 

I'd _guess_ it's because tannin sprintf only supports a subset of syntax:

> In addition to basic placeholder substitution, the following options a placeholder are implemented:
>
> Positional index
> Named arguments
> Precision / maximum width (including * precision)
> Type (only the behaviors of f, d, s, and % are applied)

Moving to that would be a breaking change because sprintf-js supports:

> A type specifier that can be any of:
> % — yields a literal % character
> b — yields an integer as a binary number
> c — yields an integer as the character with that ASCII value
> d or i — yields an integer as a signed decimal number
> e — yields a float using scientific notation
> u — yields an integer as an unsigned decimal number
> f — yields a float as is; see notes on precision above
> g — yields a float as is; see notes on precision above
> o — yields an integer as an octal number
> s — yields a string as is
> t — yields true or false
> T — yields the type of the argument[1](https://www.npmjs.com/package/sprintf-js#fn-1)
> v — yields the primitive value of the specified argument
> x — yields an integer as a hexadecimal number (lower-case)
> X — yields an integer as a hexadecimal number (upper-case)
> j — yields a JavaScript object or array as a JSON encoded string

Tannin's subset is likely sufficient and other needs can be handled in userland. This is a place where an i18n Script Module could chose to diverge without introducing a breaking change.

Reply:
> Tannin's subset is likely sufficient and other needs can be handled in userland. This is a place where an i18n Script Module could chose to diverge without introducing a breaking change.

I think what matters is feature parity with what PHP supports not what `sprints-js` can do.